packages/webassemblyjs/src/compiler/validation/stack.js

```diff
@@ -0,0 +1,444 @@
+        return false;
+      }
+      args = [moduleContext.getLocal(index)];
+      result = [moduleContext.getLocal(index)];
+      break;
+    }
+    /**
+     * get_local
+     *
+     * @see http://webassembly.github.io/spec/core/valid/instructions.html#valid-get-local
+     */
+    case "get_local": {
+      const index = instruction.args[0].value;
+      if (!moduleContext.hasLocal(index)) {
+        errors.push(`Function does not have local ${index}`);
+        return false;
+      }
+      args = [];
+      result = [moduleContext.getLocal(index)];
+      break;
+    }
+    /**
+     * block
+     *
+     * @see https://webassembly.github.io/spec/core/valid/instructions.html#valid-block
+     */
+    case "block": {
+      args = [];
+      result = instruction.result ? [instruction.result] : [];
+      break;
+    }
+    /**
+     * if
+     *
+     * @see https://webassembly.github.io/spec/core/valid/instructions.html#valid-if
+     */
+    case "if": {
+      args = ["i32"];
+      result = instruction.result ? [instruction.result] : [];
+      break;
+    }
+    /**
+     * nop
+     *
+     * @see https://webassembly.github.io/spec/core/valid/instructions.html#valid-nop
+     */
+    case "nop": {
+      args = [];
+      result = [];
+      break;
+    }
+    /**
+     * loop
+     *
+     * @see https://webassembly.github.io/spec/core/valid/instructions.html#valid-loop
+     */
+    case "loop": {
+      args = [];
+      result = instruction.resulttype ? [instruction.resulttype] : [];
+      break;
+    }
+    /**
+     * call
+     *
+     * @see https://webassembly.github.io/spec/core/valid/instructions.html#valid-call
+     */
+    case "call": {
+      if (!moduleContext.hasFunction(instruction.index.value)) {
+        errors.push(
+          `Call to undefined function index ${instruction.index.value}.`
+        );
+        return false;
+      }
+      ({ args, result } = moduleContext.getFunction(instruction.index.value));
+      break;
+    }
+    /**
+     * const
+     *
+     * @see https://webassembly.github.io/spec/core/valid/instructions.html#valid-const
+     */
+    case "const": {
+      args = [];
+      result = [instruction.object];
+      break;
+    }
+    /**
+     * drop
+     *
+     * @see https://webassembly.github.io/spec/core/valid/instructions.html#valid-const
+     */
+    case "drop": {
+      args = [ANY];
+      result = [];
+      break;
+    }
+    /**
+     * unop
+     *
+     * @see https://webassembly.github.io/spec/core/valid/instructions.html#valid-unop
+     */
+    case "clz":
+    case "ctz":
+    case "popcnt":
+    case "abs":
+    case "neg":
+    case "sqrt":
+    case "ceil":
+    case "floor":
+    case "trunc":
+    case "nearest": {
+      args = [instruction.object];
+      result = [instruction.object];
+      break;
+    }
+    /**
+     * binop
+     *
+     * @see https://webassembly.github.io/spec/core/valid/instructions.html#valid-binop
```

Comment:
It would be really cool if the validation rules could be generated from the specification - I do wonder whether that approach would be less effort in the long run?

Reply:
Interesting idea. Possibly for some that dont interact with the context (most them are already here though). The type rules that depend or modify the context (add label, change local etc) would be quite difficult I imagine

Reply:
Some rules have type variables though, so a generator would have to understand the premises and generate code to load types from the context accordingly (load type of a local for example).

We need a translation from the type rules into the moduleContext API which seems difficult to automate for me.

Also note that all modifications of the context dont happen here but in "applyInstruction", so this list is not even the full type rules in those cases.